R/featureFDRestimation.R

```diff
@@ -0,0 +1,137 @@
+#' complexFeaturesFiltered <- filterFeatures(complexFeatures,
+#'                                           min_peak_corr= 0.5,
+#'                                           min_feature_completeness= 0.5) 
+#'
+#' ## estimate the FDR
+#' estimateDecoyFDR(complexFeaturesFiltered)
+#' 
+#' #-----------------------------
+#' ## Protein Features
+#' #-----------------------------
+#' 
+#' ## Load example data
+#' proteinFeatures <-exampleProteinFeatures
+#' ## Filter feature table 
+#' proteinFeaturesFiltered <- filterFeatures(proteinFeatures,
+#'                                           min_peak_corr= 0.5,
+#'                                           min_feature_completeness= 0.5) 
+#' 
+#' ## estimate the FDR
+#' estimateDecoyFDR(proteinFeaturesFiltered)
+#' 
+
+estimateDecoyFDR <- function(features,
+                             FFT = 1,
+                             grid_search_params = NULL,
+                             verbose = TRUE){
+  
+  if("complex_id" %in% names(features)){
+    complexes <- features$complex_id
+  } else if ("protein_id" %in% names(features)) {
+    complexes <- features$protein_id
+  } else {
+    message("not a valid search result")
+  }
+  decoy_ind <- grepl("DECOY",complexes)
+  decoys <- complexes[decoy_ind]
+  targets <- complexes[!decoy_ind]
+  if(!length(decoys)>0){
+    if(verbose){
+      message("No decoy features found. Please check if decoys were available in the complex hypotheses.")
+    }
+    FDR <- 0
+  } else {
+    # estimate FDR based on detected decoys
+    FDR <- (FFT *length(decoys)) / length(targets)
+  }
+  # estimate number of true positive complexes by correcting with the estimated FDR
+  P <- length(targets)
+  TP <- length(targets)*(1-FDR)
+  D <- length(decoys)
+  
+  # output
+  if(!is.null(grid_search_params)){
+    fdr_list <- list(FDR = FDR, TP = TP, P = P, D = D)
+    param_list <- lapply(grid_search_params, function(x) as.numeric(unique(features[, x, with = F])))
+    names(param_list) <- grid_search_params
+    c(fdr_list, param_list)
+  } else {
+    list(FDR = FDR, TP = TP, P = P, D = D)
+  }
+}
+
+#' generateRandomPepTraces
+#' @description generateRandomPepTraces.
+#' @param traces traces object of type peptide
+#' @param append Logical, whether to append the generated decoy traces
+#' to the original target traces
+#' @param traces traces bject of type peptide
+#' @return traces object of type peptide with random peptides per protein
+#' @export
+#' @examples 
+#' ## Load example data
+#' exampleTraces <- examplePeptideTraces
+#' 
+#' ## Generate the decoy traces object
+#' decoyTraces <- generateRandomPepTraces(exampleTraces)
+#'                    
+#' # Inspect the result
+#' summary(decoyTraces)
+#' #--------------------
+#' ## Directly append decoy traces to target taces
+#' tracesWithDecoys <- generateRandomPepTraces(exampleTraces,
+#'                                             append = T)
+#' # Inspect the result
+#' summary(tracesWithDecoys)
+#' 
```

Comment:
I thought of doing that, but in the feature finding it generates random Hypotheses not decoy traces. So I think we can leave it the way it is for now. This function might be useful in the future though.